src/ripple/json/json_writer.h

```diff
@@ -175,6 +175,36 @@ std::string valueToQuotedString ( const char* value );
 /// \see Json::operator>>()
 std::ostream& operator<< ( std::ostream&, const Value& root );
 
+//------------------------------------------------------------------------------
+
+using write_t = std::function<void(void const*, std::size_t)>;
+
+/** Stream compact JSON to the specified function. */
+void
+stream (Json::Value const& jv, write_t write);
+
+/** Wrapper for streaming out compact json
+
+    Write
+
+        Json::Value jv;
+        out << Json::Compact{jv}
+
+    To write a single-line, compact version of `jv` to the stream, rather
+    than the typical styled version.
+*/
+class Compact
+{
+    Json::Value const & jv;
+public:
+    Compact(Json::Value const &j) : jv{j} {}
```

Comment:
So this class holds onto `jv`, which may be a temporary. this is fine if it's always used as `sink << Compact{foo()}`. However, it doesn't work if it's called as `Compact bad{foo}; sink << bad;`; Since `Json::Value` is easily movable, what do you think about moving the value `j` into `Compact` by value? This fixes the issue with temporaries.

I'm actually fine with it as-is, your call.

Reply:
I'm nervous that if we start using `Compact` with lvalues that there is a gratuitous copy of the `Json::value` just to get it printed out.  In the current code (except for the test), we only stream out rvalue `Json::value` with `Compact`, and there is no problem with that.  But in the future if we start using `Compact` with lvalues, we've created a silent performance bug.

My recommendation:

*  Change `Compact(Json::Value jv)` to `Compact(Json::Value&& jv)`.  This makes it mandatory that the argument to `Compact` is an rvalue.

*  Add a comment to this constructor that if we want to handle lvalue arguments in the future, we need to do a little more design work to handle that efficiently.

*  Change the one test that streams out an lvalue `Json::value` to stream it out as an rvalue:  `ss << Json::Compact(std::move(j));`

Now we're guaranteed efficient, and if we go non-efficient, it's a compile-time error that we can easily address in the future.  For example if we want to stream out an lvalue, we could add a `Json::value const*` member to `Compact` to point at the lvalue, and have it be `nullptr` if we're streaming an rvalue.  But no need to do that today.  We may find we never want to stream lvalues.

Reply:
I like this interface.